### ISSUE TITLE
Drastically decreasing read performence when range of keys are deleted

### ISSUE THREAD

**Issue by pcmind** [google/leveldb#610]
Performence of Iteration over a range of keys is dasticaly affected when multiple keys that share some comon prefix where previously deleted.

The use case to reproduce this issue is as follow:
1. Put a set of 1000 keys that share a common prefix.
2. Delete them all
3. For each key in the set do a search using a new DB::NewIterator (it would make more sense to use DB::get, but assume you are searching for more than one entry)
4. Repeat step 1, 2 and 3. Each loop performence hit is big.

I know that as mensioned in issue #83, commit 748539c183453bdeaff1eb0da8ccf5adacb796e7 would mitigate this issue. But as show by the folowing example, the implemented solution does not mitigate completely the issue. This is specially relevant when using LevelDB eavily with prefix searches (or mutable indexes).

I made a simple unit test in to reproduce the issue:
```cpp
TEST(DBTest, RangeDeleteAndRead) {
  do {
    for (int k = 0; k < 50; ++k) {
      Env *env = Env::Default();
      uint64_t start_micros = env->NowMicros();
      for (int i = 0; i < 2000; ++i) {
        //search for prefix
        std::string prefix = Key(i);
        std::string key1 = prefix + "-1";
        std::string key2 = prefix + "-2";
        //range search with key prefix. No entries exist but each k+1 more time is lost here
        Iterator *iter = db_->NewIterator(ReadOptions());
        iter->Seek(prefix);
        ASSERT_TRUE(!iter->Valid());
        delete iter;

        //insert values
        Put(key1, "value1");
        Put(key2, "value2");
      }
      uint64_t stop_micros = env->NowMicros();

      //delete all entries
      Iterator *iter = db_->NewIterator(ReadOptions());
      WriteBatch wb;
      iter->Seek("key");
      while (iter->Valid()) {
        if (iter->key().ToString().find("key") != 0) {
          break;
        }
        wb.Delete(iter->key().ToString());
        iter->Next();
      }
      WriteOptions wo;
      db_->Write(wo, &wb);
      delete iter;
      unsigned int us = (stop_micros - start_micros)/1000;
      fprintf(stderr,
              "Run loop %d took %d ms\n",
              k, us);
    }
  } while (ChangeOptions());
}
```

Running this test I get the following result:
```
==== Test DBTest.RangeDeleteAndRead
Run loop 0 took 78 ms
Run loop 1 took 1671 ms
Run loop 2 took 3219 ms
Run loop 3 took 4782 ms
Run loop 4 took 6312 ms
Run loop 5 took 7843 ms
...
```
The performence issue due to the fact that [db_iter](https://github.com/google/leveldb/blob/6caf73ad9dae0ee91873bcb39554537b85163770/db/db_iter.cc#L173) knows nothing about the prefix being searched by the end user.

Adding somthing like: 
```
if(!SharePrefix(&ikey)) {
  break;
}
```
to [db_iter.cc#L179|https://github.com/google/leveldb/blob/6caf73ad9dae0ee91873bcb39554537b85163770/db/db_iter.cc#L179]

**Comment by qduyang** [comment]
Hi, I may meet the same issue with frequent query, any update?

**Comment by qduyang** [comment]
Why google failed to handle those known issues in such a long time?

**Comment by philix** [comment]
@pcmind @qduyang the fix assumes the default key comparator (`memcmp`) is being used. Users can define their own comparator and define an ordering that invalidates the assumption that neighbor keys share a prefix.

**Comment by pcmind** [comment]
Yes, fix only shows that stopping early can improve greatly iteration time. 
I think that final solution should be implemented with a more generic interface, something like `iterator(fromKey, fromInclusive, toKey, toInclusive)` (because we can iterate in both directions). This would serve more use case and require no changes to Comparator.

**Comment by philix** [comment]
@pcmind totally! An internal iterator that is bounded in an inclusive key range would enable these early returns. 👏
